### PR TITLE
[CI Fest] Avoid ports clashing in integration tests

### LINKED ISSUE #67463: Integration test `test_format_avro_confluent` is flaky
Tests `test_format_avro_confluent/test.py::test_select`, `test_format_avro_confluent/test.py::test_select_auth`, `test_format_avro_confluent/test.py::test_select_auth_encoded`, `test_format_avro_confluent/test.py::test_select_auth_encoded_complex` fail in batch with

```
E               Exception: Command ['docker-compose', '--env-file', '/ClickHouse/tests/integration/test_format_avro_confluent/_instances_0/.env', '--project-name', 'roottestformatavroconfluent', '--file', '/ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_kafka.yml', '--verbose', 'up', '-d', '--renew-anon-volumes'] return non-zero code 1: compose.config.config.find: Using configuration files: /ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_kafka.yml
```

CI database

https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9mb3JtYXRfYXZyb19jb25mbHVlbnQlJwogICAgQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPiBub3coKSAtIGludGVydmFsIDYgbW9udGgKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWU=

Reports

https://s3.amazonaws.com/clickhouse-test-reports/0/5e41c29a90e50dce7cabe5a6a260a5690721dd2e/integration_tests__asan__[1_4].html
https://s3.amazonaws.com/clickhouse-test-reports/0/dfcefab9ab906c9bb42556a12923db626632b8b0/integration_tests__tsan__[5_6].html
https://s3.amazonaws.com/clickhouse-test-reports/0/823afac5aeebc5b7194c9086b0bd1430b0dfa7dc/integration_tests__asan__old_analyzer__[5_6].html

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes https://github.com/ClickHouse/ClickHouse/issues/67463. See https://github.com/ClickHouse/ClickHouse/issues/67463#issuecomment-2263421047

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 91014e2e013bf4c8c067b8a436c761d0a18a8dc5 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67619/91014e2e013bf4c8c067b8a436c761d0a18a8dc5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c0bd2120c89a4c71820e4d4d4b5d0c0a9dc66042/fast_test.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c0bd2120c89a4c71820e4d4d4b5d0c0a9dc66042/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/integration/helpers/cluster.py [hunk 1]
```diff
@@ -128,11 +128,20 @@ def run_and_check(
     return out
 
 
+# get_free_port() can return the same port multiple times.
+# To avoid it we just collect all returned ports and check for overlap.
+used_ports = []
```
Comment (nikitamikhaylov) [inline tests/integration/helpers/cluster.py:133]: Unfortunately in CI run tests in parallel within the same batch. For example for `Integration tests (release) [4/6]` we will run several `pytest` processes inside the docker container (yes, we several layers of parallelism). So, the only good solution would be just to synchronize this process via disk. I was thinking about using `sqlite`... and just specify a path for each this standalone `pytest`.

Comment (nikitamikhaylov) [reply tests/integration/helpers/cluster.py:133]: Or we can revive https://github.com/ClickHouse/ClickHouse/pull/55055 and rebind the port.

Comment (alexey-milovidov) [reply tests/integration/helpers/cluster.py:133]: Why do we need this several levels of parallelism? Maybe run separate Docker containers for each test?

Comment (Avogar) [reply tests/integration/helpers/cluster.py:133]: Okey, probably I found case when there was a clash between parallel runs: https://s3.amazonaws.com/clickhouse-test-reports/0/aabf7a576ce42ef03456bf846784adfd1635ccaa/integration_tests__release__[3_4]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel5_0.log.
Let's revive rebinding indeed, will add in this PR

Comment (nikitamikhaylov) [reply tests/integration/helpers/cluster.py:133]: I've assigned port rebinding to the @fm4v already...

Comment (Avogar) [reply tests/integration/helpers/cluster.py:133]: No problem, then I will pick something else. We had a clash on fixing ports clashing :)

**Pull request closed without merging**